sys/drv/pit.c

```diff
@@ -12,48 +12,67 @@
 #define PIIX4_IDE_VENDOR_ID 0x8086
 #define PIIX4_IDE_DEVICE_ID 0x7111
 
+typedef union {
+  /* assumes little endian order */
+  struct {
+    uint32_t lo;
+    uint32_t hi;
+  };
+  uint64_t val;
+} counter_t;
+
 typedef struct pit_state {
   resource_t *regs;
-  spin_t lock;
   intr_handler_t intr_handler;
   timer_t timer;
-  uint64_t period_frac;        /* period as a fraction of a second */
-  uint16_t period_cntr;        /* period as PIT counter value */
-  volatile uint16_t last_cntr; /* last seen counter value */
-  volatile bool overflow;      /* counter overflow detected? */
-  bintime_t time;              /* last time measured by the timer */
+  uint16_t period_cntr; /* period as PIT counter value */
 } pit_state_t;
 
 #define inb(addr) bus_read_1(pit->regs, (addr))
 #define outb(addr, val) bus_write_1(pit->regs, (addr), (val))
 
 static void pit_set_frequency(pit_state_t *pit, uint16_t period) {
-  assert(spin_owned(&pit->lock));
-
   outb(TIMER_MODE, TIMER_SEL0 | TIMER_16BIT | TIMER_RATEGEN);
   outb(TIMER_CNTR0, period & 0xff);
   outb(TIMER_CNTR0, period >> 8);
   pit->period_cntr = period;
 }
 
-static uint16_t pit_get_counter(pit_state_t *pit) {
-  assert(spin_owned(&pit->lock));
-
+static uint16_t pit_get_counter16(pit_state_t *pit) {
+  uint16_t count = 0;
   outb(TIMER_MODE, TIMER_SEL0 | TIMER_LATCH);
-  uint16_t value = 0;
-  value |= inb(TIMER_CNTR0);
-  value |= inb(TIMER_CNTR0) << 8;
-  return pit->period_cntr - value;
+  count |= inb(TIMER_CNTR0);
+  count |= inb(TIMER_CNTR0) << 8;
+  return count;
+}
+
+static uint64_t pit_get_counter64(pit_state_t *pit) {
+  static uint16_t counter16_last = 0;
```

Comment:
Using global state that is not tied to `pit_state_t` prevents this driver to be instantiated twice. While this is not a problem, it's considered a bad practice while implementing device drivers. Please move `counter16_last` and `counter64_last` to `pit_state_t`.

Reply:
Done.